### PR TITLE
Disallow duplicate CTE aliases

### PR DESCRIPTION
**Pull request by yashmayya**
- Currently, a query like `WITH tmp AS (SELECT * FROM a LIMIT 1), tmp AS (SELECT * FROM a LIMIT 2) SELECT * FROM tmp` doesn't return an error and runs fine.
- By default, Calcite just uses the last CTE for the alias in such cases. However, this can lead to confusing and unexpected results in case there was a human error while writing a complex query with many CTEs.
- In databases like Postgres and MySQL too, such queries return a clear error.
- Postgres:
```
ERROR:  WITH query name "tmp" specified more than once
```
- MySQL:
```
ERROR 1066 (42000): Not unique table/alias: 'tmp'
```
- Pinot (after this patch):
```
Duplicate alias in WITH: 'tmp' at line 1, column 40
```

### CODE REVIEW

File: pinot-query-planner/src/main/java/org/apache/pinot/query/validate/Validator.java [hunk 1]
```diff
@@ -147,4 +151,20 @@ protected void addToSelectList(List<SqlNode> list, Set<String> aliases,
...
   }
+
+  @Override
+  public void validateWith(SqlWith with, SqlValidatorScope scope) {
+    Set<String> withNames = new HashSet<>();
+
+    for (SqlNode sqlWithItem : with.withList) {
+      for (String name : ((SqlWithItem) sqlWithItem).name.names) {
+        if (withNames.contains(name)) {
+          throw new RuntimeException("Duplicate alias in WITH: '" + name + "'");
```
Comment (gortiz) [inline pinot-query-planner/src/main/java/org/apache/pinot/query/validate/Validator.java:162]: Can we also provide lines? Given we are at SqlNode here, we should be able to access them

File: pinot-query-planner/src/test/java/org/apache/pinot/query/QueryCompilationTest.java [hunk 2]
```diff
@@ -385,6 +385,12 @@ public void testGetTableNamesForQuery() {
     assertEquals(tableNames.get(0), "a");
   }
 
+  @Test(expectedExceptions = RuntimeException.class)
```
Comment (gortiz) [inline pinot-query-planner/src/test/java/org/apache/pinot/query/QueryCompilationTest.java:388]: can we also verify the message is what we expect?

File: pinot-query-planner/src/test/java/org/apache/pinot/query/QueryCompilationTest.java [hunk 3]
```diff
@@ -385,10 +385,15 @@ public void testGetTableNamesForQuery() {
...
+  @Test
   public void testDuplicateWithAlias() {
     String query = "WITH tmp AS (SELECT * FROM a LIMIT 1), tmp AS (SELECT * FROM a LIMIT 2) SELECT * FROM tmp";
-    _queryEnvironment.getTableNamesForQuery(query);
+    try {
+      _queryEnvironment.getTableNamesForQuery(query);
+      fail("Expected compilation of query with duplicate CTE alias to throw an exception");
+    } catch (RuntimeException e) {
+      assertTrue(e.getCause().getMessage().contains("Duplicate alias in WITH: 'tmp'"));
+    }
```
Comment (gortiz) [inline pinot-query-planner/src/test/java/org/apache/pinot/query/QueryCompilationTest.java:396]: nit: You can use `@Test(expectedExceptionsMessageRegExp = "regexp")` which will give you a better error message

File: pinot-query-planner/src/test/java/org/apache/pinot/query/QueryCompilationTest.java (hunk 2 above)
Comment (yashmayya) [reply pinot-query-planner/src/test/java/org/apache/pinot/query/QueryCompilationTest.java]: I'd initially tried that with `expectedExceptionsMessageRegExp` but looks like that only checks the top level exception message (and not nested exceptions) which in our case is the `RuntimeException` with `Error composing query plan for:...` from [here](https://github.com/apache/pinot/blob/205a75debbdd2c161f6f830a71e467605fac6ec4/pinot-query-planner/src/main/java/org/apache/pinot/query/QueryEnvironment.java#L140). I'm not sure why we use TestNG and not JUnit which is a lot more feature rich 🤷 

I updated the test to use a more old school style of catching the expected exception and making assertions in the catch block.

File: pinot-query-planner/src/test/java/org/apache/pinot/query/QueryCompilationTest.java (hunk 3 above)
Comment (yashmayya) [reply pinot-query-planner/src/test/java/org/apache/pinot/query/QueryCompilationTest.java:396]: Answered [here](https://github.com/apache/pinot/pull/13664#discussion_r1684468878).

**Review by gortiz — APPROVED** [review]

**Pull request merged**